### ISSUE TITLE
Missing CUBICSPLINE interpolation mode

### ISSUE THREAD

**Issue by kungfooman** [playcanvas/playcanvas-gltf#43]
https://youtu.be/H1nRE2ipqQU

[![animation bug](https://img.youtube.com/vi/H1nRE2ipqQU/0.jpg)](https://www.youtube.com/watch?v=H1nRE2ipqQU)

I tested threejs, babylon and win10 "3d builder", they all show the same animation (still buggy animation, but I guess thats the fault of the khronos blender exporter), but something in playcanvas-gltf animation is definitive more wrong

I uploaded door.gltf, door.bin and door.blend here when anybody wants to use this as simple/small reference bug: http://killtube.org/downloads/gltf/door/ (asset is public domain, i just made it for debugging all the gltf animation problems)

**Comment by kungfooman** [comment]
I kinda fixed the animation by hand now, to first figure out whats wrong exactly and to test if it works with proper values... and it works. So the real problem is some data fetching/generating for the pos/scale animKeys.

This is how the animation looks fixed in playcanvas-gltf:

![fixed_keys](https://user-images.githubusercontent.com/5236548/46953224-9135fa00-d08d-11e8-95f8-7b9aa9d2fc2c.gif)

```JS
// for some reason khronos only exports 4 keyframes, but in blender i have 5
// this adds the very first keyframe (time == 0.0)
// just set position to dummy value of [0,0,0], updating correct position in next code
viewer.clip.animCurves[0].insertKey(AnimationKeyableType.VEC, 0.0, new pc.Vec3(0,0,0))

// from blender to playcanvas, swizzle xyz = xzy
// i simply hovered over the keyframes in blender and copy/pasted the values
viewer.clip.animCurves[0].animKeys[0].value.set( -1.30048 , 0.087398, 0.12464  )
viewer.clip.animCurves[0].animKeys[1].value.set( -1.13242 , 1.90772 , 0.120238 )
viewer.clip.animCurves[0].animKeys[2].value.set( -0.033788, 2.62254 , 0.038914 )
viewer.clip.animCurves[0].animKeys[3].value.set(  1.17403 , 2.12943 , 0.022512 )
viewer.clip.animCurves[0].animKeys[4].value.set(  1.3388  , 0.29553 , 0.207579 )

// scale is 1.258 for every animkey
// the exporter only exports 4 animKeys aswell, but a proper exporter should have 5 too
// i just done care because its all the same value atm
viewer.clip.animCurves[2].animKeys[0].value.set( 1.258  , 1.258 , 1.258 )
viewer.clip.animCurves[2].animKeys[1].value.set( 1.258  , 1.258 , 1.258 )
viewer.clip.animCurves[2].animKeys[2].value.set( 1.258  , 1.258 , 1.258 )
viewer.clip.animCurves[2].animKeys[3].value.set( 1.258  , 1.258 , 1.258 )
```

The buggy default values for positioning:

```JS
viewer.clip.animCurves[0].animKeys.map((animKey) => animKey.value)

0: Vec3 {x: 0, y: 0, z: 0}
1: Vec3 {x: -1.3004825115203857, y: 0.08739840984344482, z: -0.12463954091072083}
2: Vec3 {x: 0, y: 0, z: 0}
3: Vec3 {x: 1.2100780010223389, y: 3.5630946159362793, z: 0.03169051557779312}
```

The buggy default values for scaling:

```JS
viewer.clip.animCurves[2].animKeys.map((animKey) => animKey.value)

0: Vec3 {x: 0, y: 0, z: 0}
1: Vec3 {x: 1.2582228183746338, y: 1.2582228183746338, z: 1.2582228183746338}
2: Vec3 {x: 0, y: 0, z: 0}
3: Vec3 {x: 0, y: 0, z: 0}
```

**Comment by willeastcott** [comment]
Hey @thomasbiang, do you know what's happening here?

**Comment by kungfooman** [comment]
I slowly get a grasp, seems like CUBICSPLINE interpolation mode just isn't implemented.

https://github.com/playcanvas/playcanvas-gltf/blob/c198f289111e0fa19d1e7a5cf89b0e6f7968f471/src/playcanvas-gltf.js#L143-L176

For the sake of simple debugging I just added this (would be handy to add a debug option to enable this, so the release version wouldn't waste memory/referenced object):

```JS
curve.sampler = sampler;
curve.times = times;
curve.values = values;
clip.addCurve(curve);
```

and selected the buggy curve:

![cubicspline](https://user-images.githubusercontent.com/5236548/46965344-ea605680-d0aa-11e8-86ad-fb6fd4da5d64.png)

Float32Array with 36 elements, blender exported 4 animKeys, so each CUBICSPLINE uses 9 floats. Googled a bit and it makes quite sense from the specification:

https://github.com/KhronosGroup/glTF/tree/master/specification/2.0

> The animation's interpolation is computed using a cubic spline with specified tangents. The number of output elements must equal three times the number of input elements. For each input element, the output stores three elements, an in-tangent, a spline vertex, and an out-tangent. There must be at least two keyframes when using this interpolation.

However, reading the tangent/spline data seems to be a little artform in itself, e.g. this is how threejs does it:

https://github.com/mrdoob/three.js/blob/2410ad1a8f5844df00191600eafda59def537830/examples/js/loaders/GLTFLoader.js#L1005-L1016

**Comment by thomasbiang** [comment]
We do have CUBIC interpolation implemented. I am looking into it.

**Comment by kungfooman** [comment]
Thanks for looking into this. I also figured out now why the first animKey was dropped in the Blender exporter, so this is the last issue for some proper Blender/NLA export pipeline (at least for my needs currently).

Blender first-animKey-dropped exporter issue: https://github.com/KhronosGroup/glTF-Blender-Exporter/issues/322

Fixed door with gltf/bin/blend: http://killtube.org/downloads/gltf/door_all_animkeys/

**Comment by willeastcott** [comment]
Fixed in #45
